Add support for importing C++ inline functions

bricknerb opened the pull request:
This requires:
* Making `FunctionDecl` mutable since generating code (`HandleTopLevelDecl()`) requires a mutable declaration and since we manually add `used` attribute to force code generation.
* Passing the file system to `Lower` since it's needed by Clang code generation.
* Creating an internal Clang LLVM module and link it against the Carbon LLVM module.

Demo:

```c++
// hello_world.h

extern int puts;
inline void hello_world() {
  ((int (*)(const char*))&puts)("hello world");
}
```

```carbon
// main.carbon

library "Main";

import Cpp library "hello_world.h";

fn Run() -> i32 {
  Cpp.hello_world();
  return 0;
}
```

```shell
$ bazel-bin/toolchain/carbon compile main.carbon
$ bazel-bin/toolchain/carbon link main.o --output=demo
$ ./demo
hello world
```

Based on https://github.com/carbon-language/carbon-lang/pull/5406.

Part of #5405.

bricknerb committed:
Add C++ interop inline function tests
This shows that `inline` is ignored and the function definition is not generated.

bricknerb committed:
Add a TODO to generate the definition of the inline function.

bricknerb committed:
Add support for importing C++ inline functions
This requires:
* Making `FunctionDecl` mutable since generating code (`HandleTopLevelDecl()`) requires a mutable declaration and since we manually add `used` attribute to force code generation.
* Passing the file system to `Lower` since it's needed by Clang code generation.
* Creating an internal Clang LLVM module and link it against the Carbon LLVM module.

bricknerb committed:
Instead of adding the `used` attribute, call `GetAddrOfGlobal()` to create the llvm function so it will be generated.

zygoloid commented on toolchain/lower/file_context.h:
```diff
@@ -45,6 +49,13 @@ class FileContext {
 
+  // Creates the Clang `CodeGenerator` to generate LLVM module from imported C++
+  // code. Returns null when not importing C++. Can be called only after the
+  // following data members are initialized: `llvm_context_`, `llvm_module_`,
+  // `fs_`, `cpp_ast_`, `cpp_header_search_options_`,
+  // `cpp_preprocessor_options_`, `cpp_code_gen_options_`, .
```
The last part of this comment seems a bit redundant, given that most of those members are initialized by the constructor. Also if there's a precondition here, it'd be better for it to be user-facing -- that is, saying which member functions of this class must be called first, not which data members must be initialized.

bricknerb replied:
The reason I did it this way was because `CreateCppCodeGenerator()` is called within the constructor, when initializing `cpp_code_generator_`.
I could create it in `Init()` instead, before it is initialized.
WDYT?

zygoloid replied:
I see. Then maybe you could add to the comment that this function is called from the constructor while still initializing some of the class members? That would have helped me understand what's going on.

Alternatively, could you call this from the constructor body rather than from the member initializers, so it's only called once the invariants of the class are already set up?

bricknerb replied:
Thanks!
Calling the constructor body.

zygoloid commented on toolchain/lower/file_context.cpp:
```diff
@@ -390,6 +423,24 @@ auto FileContext::BuildFunctionBody(SemIR::FunctionId function_id,
+                                             /*isForDefinition=*/false));
+
+    // Emit the function code.
+    cpp_code_generator_->HandleTopLevelDecl(clang::DeclGroupRef(cpp_def));
```
If this call is necessary, are we able to handle the case where an inline function calls another inline function? Eg:
```c++
// --- inline_function_with_call.h

inline void bar() {}
inline void foo() { bar(); }
```
```carbon
// --- import_inline_function_decl.carbon

library "[[@TEST_NAME]]";

import Cpp library "inline_function_decl.h";

fn MyF() {
  // Should cause both `foo` and `bar` to be emitted.
  Cpp.foo();
}
```

bricknerb replied:
No, see [Discord note](https://discord.com/channels/655572317891461132/768530752592805919/1370038287967981579)
Merged with base, added an explicit test for that with TODOs in the test and in the code.

zygoloid commented on toolchain/lower/file_context.cpp:
```diff
@@ -390,6 +423,24 @@ auto FileContext::BuildFunctionBody(SemIR::FunctionId function_id,
+    // that code generation (`CodeGenModule::EmitGlobal()`) would see this
+    // function name (`CodeGenModule::getMangledName()`), and will generate its
+    // definition.
+    CARBON_DCHECK(
+        cpp_code_generator_->GetAddrOfGlobal(clang::GlobalDecl(cpp_def),
+                                             /*isForDefinition=*/false));
```
Please don't put code with side-effects inside a CHECK or especially a DCHECK. Instead, call the function outside the CHECK and separately test its result.

bricknerb replied:
Makes sense, thanks, done.

zygoloid commented on toolchain/lower/file_context.cpp:
```diff
@@ -118,6 +127,14 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
 
+  if (cpp_code_generator_) {
+    cpp_code_generator_->HandleTranslationUnit(cpp_ast()->getASTContext());
+    CARBON_CHECK(!llvm::Linker::linkModules(
+        /*Dest=*/*llvm_module_,
+        /*Src=*/std::unique_ptr<llvm::Module>(
+            cpp_code_generator_->ReleaseModule())));
```
(Also here: please don't put code with side effects inside a CHECK.)

bricknerb replied:
Thanks, done.
Is this because the side effect might run more than once on failure?

bricknerb committed:
Call `GetAddrOfGlobal()` outside `CARBON_DCHECK()`

bricknerb committed:
Move the call to `linkModules()` outside `CARBON_CHECK()`

bricknerb committed:
Merge branch 'main' into inline5

bricknerb committed:
Remove `inline_function_decl` in `toolchain/check/testdata/interop/cpp/no_prelude/function_decl.carbon` as it already exists in `toolchain/check/testdata/interop/cpp/no_prelude/function_decl_inline.carbon`

bricknerb committed:
Added `todo_import_inline_recursive_function_decl` test

bricknerb committed:
Add add `TODO` clarifying how we should support inline functions based on https://discord.com/channels/655572317891461132/768530752592805919/1370509111585935443

bricknerb committed:
retrigger checks

bricknerb committed:
Clear target triple and data layout to make tests pass

bricknerb committed:
Revert "Clear target triple and data layout to make tests pass"

This reverts commit 8f584efd9313abfe83560fc4551a3462a3332e60.

bricknerb commented:
FTR, I'm trying to fix the tests, which seem to fail on the difference between different targets and (perhaps as a result) debug information in the module.
I can manually remove the target but the impact on the debug information remains and still haven't found the option to fix this.
Let me know if you know how to fix this (still haven't debugged this fully).

bricknerb committed:
Propagate Carbon's `target` flag to `buildASTFromCodeWithArgs()` since it impacts how the LLVM module is generated.
Set it to a constant value in `toolchain/lower/testdata/interop/cpp/function_decl.carbon`

bricknerb commented:
> FTR, I'm trying to fix the tests, which seem to fail on the difference between different targets and (perhaps as a result) debug information in the module. I can manually remove the target but the impact on the debug information remains and still haven't found the option to fix this. Let me know if you know how to fix this (still haven't debugged this fully).

I'm now configuring the `target` when creating the ASTUnit.
Let me know what you think.

bricknerb committed:
Merge branch 'main' into inline5

zygoloid commented on toolchain/check/import_cpp.cpp:
```diff
@@ -179,7 +179,8 @@ class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
 static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
                         llvm::ArrayRef<Parse::Tree::PackagingNames> imports,
-                        llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs)
+                        llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs,
```
This seems OK for now, but I'm worried that we're going to end up plumbing a lot of clang configuration information through check. Let's talk in the interop meeting about options here.

bricknerb replied:
As discussed in the meeting, we're leaving this as is for now and reevaluate when we use pcms (modules).

zygoloid reviewed (approved):
Thanks, looks good other than a trivial comment about a comment :)

bricknerb committed:
Move the call to `CreateCppCodeGenerator()` to the constructor body.
